sdk/src/main/scala/com/horizen/account/state/AccountStateView.scala

```diff
@@ -151,25 +151,43 @@ class AccountStateView(
     // Tx context for stateDB, to know where to keep EvmLogs
     setupTxContext(txHash, txIndex)
 
-    // apply message to state
-    val status = try {
-      applyMessage(msg, blockGasPool)
-      ReceiptStatus.SUCCESSFUL
-    } catch {
-      // any other exception will bubble up and invalidate the block
-      case err: ExecutionFailedException =>
-        log.error(s"applying message failed, tx.id=${ethTx.id}", err)
-        ReceiptStatus.FAILED
-    } finally {
-      // finalize pending changes, clear the journal and reset refund counter
+    val revisionId = snapshot
+    val initialBlockGas = blockGasPool.getGas
+    try {
+      // apply message to state
+      val status = try {
+        applyMessage(msg, blockGasPool)
+        ReceiptStatus.SUCCESSFUL
+      } catch {
+        // any other exception will bubble up and invalidate the block
+        case err: ExecutionFailedException =>
+          log.error(s"applying message failed, tx.id=${ethTx.id}", err)
+          ReceiptStatus.FAILED
+      } finally {
+        // finalize pending changes, clear the journal and reset refund counter
+        // make sure we disable automatic gas consumption in case a message processor enabled it
+        disableGasTracking()
+      }
+      val consensusDataReceipt = new EthereumConsensusDataReceipt(
+        ethTx.version(), status.id, blockGasPool.getUsedGas, getLogs(txHash))
+      log.debug(s"Returning consensus data receipt: ${consensusDataReceipt.toString()}")
+      consensusDataReceipt
+    }
+    catch {
+      case e : Exception =>
```

Comment:
This exception case could be moved next to the other. All the exception will be coming from `applyMessage` anyway.

Reply:
Instead of wrapping the whole code base, I propose to add one more parameter to the `applyTransaction` methods: `finalizeChanges: Boolean` that has default value `true`.
Forger will:
1) create a snapshot
2) execute`applyTransaction` with `finalizeChanges=false`
3) in case applied tx doesn't feet the block -> do revertToSnapshot. Otherwise, do `stateDb.finalizeChanges()`

Reply:
@JcBernack: This fix was made because there was a bug in the EthereumConsensusDataReceipt.toString method that caused a NullPointerException. Because of that exception, the tx was not included in the block by the forger but the state wasn't reverted, so the rootstatehash in the block was wrong etc. etc. Now, the root cause of the NullPointerException is fixed and I agree with you that, except for bugs, there are no real possible causes of unexpected exceptions here. Still I think that it is better to be safe than sorry and to protect all the code. Consider also that maybe in the future new, more "dangerous" line of codes could be added in the same place and maybe one could forget to protect them.

Reply:
@i-Alex: I thought that it was a responsibility of AccountStateView class to guarantee that the state remained consistent, regardless of what the caller would have done next. But if you think it is better to leave the forger to call finalizeChanges, it's ok for me.